File: android/app/src/main/java/app/notomorrow/feature/workout/WorkoutMiniBar.kt
package app.notomorrow.feature.workout

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.rotate
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.onClick
import androidx.compose.ui.semantics.role
import androidx.compose.ui.semantics.stateDescription
import androidx.compose.ui.unit.dp
import app.notomorrow.designsystem.CapsuleShape
import app.notomorrow.designsystem.GlassStyle
import app.notomorrow.designsystem.NT
import app.notomorrow.designsystem.NtIcon
import app.notomorrow.designsystem.NtIcons
import app.notomorrow.designsystem.NtText
import app.notomorrow.designsystem.ProgressRing
import app.notomorrow.designsystem.liquidGlass
import app.notomorrow.designsystem.ntPlainClickable
import app.notomorrow.designsystem.pressScale
import app.notomorrow.designsystem.rememberSecondTicker
import app.notomorrow.designsystem.sfIconSize
import app.notomorrow.designsystem.tabular
import app.notomorrow.rest.RestTimerState
import app.notomorrow.util.Fmt
import app.notomorrow.util.S
import app.notomorrow.util.rememberNtStrings

/**
 * What the mini bar says at one moment — 1:1 port of `WorkoutMiniBarState`
 * (`WorkoutMiniBar.swift`). Pure, so the phrasing rules are testable.
 *
 * @property elapsed "42:10" / "1:02:03" since the workout started.
 * @property restRemaining seconds of rest left, `null` when not resting.
 * @property restFraction share of the rest still to go (1 → 0), for the ring.
 * @property exerciseName resting: the exercise up next; otherwise the one the user is on. Empty
 *   when there is none.
 */
data class WorkoutMiniBarState(
    val elapsed: String,
    val restRemaining: Double?,
    val restFraction: Double,
    val exerciseName: String,
) {
    val isResting: Boolean get() = restRemaining != null

    /**
     * TalkBack's state for the bar: the exercise, or "Rest, Squat" while resting. The countdown
     * stays out on purpose: a focused node whose state changes is read out again, and a
     * second-by-second countdown would make TalkBack repeat the bar every tick. What changes at
     * most once a minute (the elapsed minutes in the label) is coarse enough to follow.
     */
    fun accessibilityValue(restLabel: String): String {
        if (!isResting) return exerciseName
        return if (exerciseName.isEmpty()) restLabel else "$restLabel, $exerciseName"
    }

    companion object {
        fun of(
            startedAt: Long,
            now: Long,
            restEnd: Long?,
            restTotal: Int,
            upNextName: String,
            currentName: String?,
        ): WorkoutMiniBarState {
            val elapsed = Fmt.elapsed((now - startedAt) / 1000.0)
            if (restEnd != null && restEnd > now) {
                val remaining = (restEnd - now) / 1000.0
                return WorkoutMiniBarState(
                    elapsed = elapsed,
                    restRemaining = remaining,
                    restFraction = if (restTotal > 0) minOf(1.0, remaining / restTotal) else 0.0,
                    exerciseName = upNextName.ifEmpty { currentName.orEmpty() },
                )
            }
            return WorkoutMiniBarState(
                elapsed = elapsed,
                restRemaining = null,
                restFraction = 0.0,
                exerciseName = currentName.orEmpty(),
            )
        }
    }
}

/**
 * The workout in progress, pinned above the tab bar on every tab — 1:1 port of `WorkoutMiniBar`
 * (`WorkoutMiniBar.swift`): name, elapsed time and current exercise; while resting, the
 * countdown, what is up next and a Skip button. Tapping anywhere else expands the workout.
 *
 * Stateless: everything comes from the workout's view model and the rest timer, so collapsing and
 * tab switches lose nothing. A Liquid Glass capsule like the tab bar, and like it drawn **outside**
 * the recorded page (`MainTabScaffold`); its own graphics layer keeps the one-second tick from
 * repainting anything else.
 */
@Composable
fun WorkoutMiniBar(
    name: String,
    startedAt: Long,
    currentExerciseName: String?,
    rest: RestTimerState,
    onExpand: () -> Unit,
    onSkipRest: () -> Unit,
    modifier: Modifier = Modifier,
) {
    // Not `rest.isRunning()` at composition: the countdown and the resting look flip on the tick.
    val now by rememberSecondTicker()
    val strings = rememberNtStrings()
    val state = WorkoutMiniBarState.of(
        startedAt = startedAt,
        now = now,
        restEnd = rest.endAt,
        restTotal = rest.totalSeconds,
        upNextName = rest.exerciseName,
        currentName = currentExerciseName,
    )
    val restLabel = stringResource(S.timer_rest)
    val label = stringResource(
        S.workout_miniBar_label_s_s,
        name,
        Fmt.duration(((now - startedAt).coerceAtLeast(0L)) / 1000.0, strings),
    )
    val value = state.accessibilityValue(restLabel)
    // TalkBack reads "Double-tap to <action>", so this is the action phrase ("otworzyć trening"),
    // not the iOS hint sentence ("Otwiera trening").
    val action = stringResource(S.workout_miniBar_action)

    Row(
        modifier = modifier
            .fillMaxWidth()
            .height(WorkoutMiniBarTokens.height)
            .graphicsLayer()
            .liquidGlass(CapsuleShape, GlassStyle.Regular)
            .padding(start = 12.dp, end = if (state.isResting) 8.dp else 12.dp),
        horizontalArrangement = Arrangement.spacedBy(8.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Row(
            modifier = Modifier
                .weight(1f)
                .fillMaxHeight()
                .ntPlainClickable(role = Role.Button, onClickLabel = action, onClick = onExpand)
                .clearAndSetSemantics {
                    contentDescription = label
                    stateDescription = value
                    role = Role.Button
                    onClick(label = action) {
                        onExpand()
                        true
                    }
                },
            horizontalArrangement = Arrangement.spacedBy(12.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Leading(state)
            Column(Modifier.weight(1f), verticalArrangement = Arrangement.spacedBy(1.dp)) {
                NtText(text = name, style = NT.Fonts.subheadlineBold, color = NT.Colors.ink, maxLines = 1)
                Detail(state, restLabel)
            }
            if (!state.isResting) {
                // `chevron.up`: the set has only the down chevron, turned over.
                NtIcon(
                    icon = NtIcons.ChevronDown,
                    modifier = Modifier.size(24.dp).rotate(180f),
                    size = sfIconSize(13f),
                    tint = NT.Colors.ink2,
                )
            }
        }
        if (state.isResting) SkipButton(onClick = onSkipRest)
    }
}

/** An 8 dp ember dot, or while resting a 24 dp ring draining with the rest. */
@Composable
private fun Leading(state: WorkoutMiniBarState) {
    Box(Modifier.size(24.dp), contentAlignment = Alignment.Center) {
        if (state.isResting) {
            ProgressRing(
                progress = state.restFraction,
                modifier = Modifier.size(24.dp),
                lineWidth = 3.dp,
            )
        } else {
            Box(Modifier.size(8.dp).background(NT.Colors.ember, CircleShape))
        }
    }
}

/** "42:10 · Bench press", or while resting "Rest 1:12 · Squat" with the countdown in ember. */
@Composable
private fun Detail(state: WorkoutMiniBarState, restLabel: String) {
    val style = NT.Fonts.footnote.tabular()
    Row(verticalAlignment = Alignment.CenterVertically) {
        val remaining = state.restRemaining
        if (remaining != null) {
            NtText(text = "$restLabel ", style = style, color = NT.Colors.ink2, maxLines = 1)
            NtText(text = Fmt.clock(remaining), style = style, color = NT.Colors.ember, maxLines = 1)
        } else {
            NtText(text = state.elapsed, style = style, color = NT.Colors.ink2, maxLines = 1)
        }
        if (state.exerciseName.isNotEmpty()) {
            NtText(
                text = " · " + state.exerciseName,
                modifier = Modifier.weight(1f, fill = false),
                style = style,
                color = NT.Colors.ink2,
                maxLines = 1,
            )
        }
    }
}

/** `common.skip` on an ink capsule, 32 dp tall in a 44 dp hit area. */
@Composable
private fun SkipButton(onClick: () -> Unit) {
    Box(
        modifier = Modifier.heightIn(min = NT.Size.control).pressScale(onClick = onClick),
        contentAlignment = Alignment.Center,
    ) {
        Box(
            modifier = Modifier
                .height(32.dp)
                .background(NT.Colors.ink, CapsuleShape)
                .padding(horizontal = 14.dp),
            contentAlignment = Alignment.Center,
        ) {
            NtText(
                text = stringResource(S.common_skip),
                style = NT.Fonts.footnoteBold,
                color = NT.Colors.onPrimary,
                maxLines = 1,
            )
        }
    }
}

/** The iOS bar's geometry (`WorkoutMiniBar.height`, its inset padding), in one place. */
object WorkoutMiniBarTokens {
    /** `NT.Size.cardButton`. */
    val height = NT.Size.cardButton

    /** Clear space between the bar and the tab bar under it. */
    val gap = 8.dp

    /** Everything the bar adds to the bottom chrome while it shows. */
    val reserved = height + gap
}

File: NoTomorrow/Features/Workout/WorkoutMiniBar.swift
import SwiftUI
import SwiftData

/// What the mini bar says at one moment. Pure, so the phrasing rules are testable.
struct WorkoutMiniBarState: Equatable {
    /// "42:10" / "1:02:03" since the workout started.
    var elapsed: String
    /// Seconds of rest left, nil when not resting.
    var restRemaining: TimeInterval?
    /// Share of the rest still to go (1 → 0), for the ring.
    var restFraction: Double
    /// Resting: the exercise up next. Otherwise: the exercise the user is on. Empty when there is none.
    var exerciseName: String

    var isResting: Bool { restRemaining != nil }

    init(startedAt: Date, now: Date, restEnd: Date?, restTotal: Int, upNextName: String, currentName: String?) {
        elapsed = Fmt.elapsed(now.timeIntervalSince(startedAt))
        if let restEnd, restEnd > now {
            let remaining = restEnd.timeIntervalSince(now)
            restRemaining = remaining
            restFraction = restTotal > 0 ? min(1, remaining / Double(restTotal)) : 0
            exerciseName = upNextName.isEmpty ? (currentName ?? "") : upNextName
        } else {
            restRemaining = nil
            restFraction = 0
            exerciseName = currentName ?? ""
        }
    }
}

/// The workout in progress, pinned above the tab bar on every tab: name, elapsed time and current exercise;
/// while resting, the countdown, what is up next and a Skip button. Tapping anywhere else expands the workout.
/// Stateless: everything comes from the session and the rest timer, so collapsing and tab switches lose nothing.
struct WorkoutMiniBar: View {
    @Environment(WorkoutSessionController.self) private var session
    @Environment(RestTimerController.self) private var restTimer
    @Environment(\.modelContext) private var context

    var body: some View {
        if let workout = session.workout(in: context), workout.isActive {
            // Not `restTimer.isRunning`: that only flips when something calls `finishIfElapsed()`.
            TimelineView(.periodic(from: .now, by: 1)) { ctx in
                bar(workout, state: state(for: workout, now: ctx.date), now: ctx.date)
            }
        }
    }

    private func state(for workout: Workout, now: Date) -> WorkoutMiniBarState {
        let current = session.model?.currentExercise
            ?? ActiveWorkoutModel.currentExercise(in: workout.sortedExercises, expandedID: nil)
        return WorkoutMiniBarState(startedAt: workout.startedAt, now: now,
                                   restEnd: restTimer.endDate, restTotal: restTimer.totalSeconds,
                                   upNextName: restTimer.exerciseName,
                                   currentName: current?.exercise?.localizedName)
    }

    private func bar(_ workout: Workout, state: WorkoutMiniBarState, now: Date) -> some View {
        HStack(spacing: 8) {
            Button { session.expand() } label: {
                HStack(spacing: 12) {
                    leading(state)
                    VStack(alignment: .leading, spacing: 1) {
                        Text(WorkoutStrings.displayName(workout.name))
                            .font(NT.Fonts.subheadlineBold).foregroundStyle(NT.Colors.ink).lineLimit(1)
                        detail(state)
                    }
                    Spacer(minLength: 0)
                    if !state.isResting {
                        Image(systemName: "chevron.up")
                            .font(.system(size: 13, weight: .bold))
                            .foregroundStyle(NT.Colors.ink2)
                            .frame(width: 24, height: 24)
                    }
                }
                .frame(maxHeight: .infinity)
                .contentShape(Rectangle())
            }
            .buttonStyle(.plain)
            .accessibilityLabel(Text("workout.miniBar.label \(WorkoutStrings.displayName(workout.name)) \(Fmt.duration(now.timeIntervalSince(workout.startedAt)))"))
            .accessibilityValue(accessibilityValue(state))
            .accessibilityHint(Text("workout.miniBar.hint"))

            if state.isResting {
                Button { restTimer.skip() } label: {
                    Text("common.skip")
                        .font(NT.Fonts.footnoteBold).foregroundStyle(NT.Colors.onPrimary)
                        .padding(.horizontal, 14)
                        .frame(height: 32)
                        .background(NT.Colors.ink, in: Capsule())
                        .frame(minHeight: NT.Size.control)
                        .contentShape(Rectangle())
                }
                .buttonStyle(PressScale())
            }
        }
        .padding(.leading, 12)
        .padding(.trailing, state.isResting ? 8 : 12)
        .frame(height: WorkoutMiniBar.height)
        .miniBarChrome()
    }

    @ViewBuilder
    private func leading(_ state: WorkoutMiniBarState) -> some View {
        if state.isResting {
            ProgressRing(progress: state.restFraction, lineWidth: 3)
                .frame(width: 24, height: 24)
        } else {
            Circle().fill(NT.Colors.ember)
                .frame(width: 8, height: 8)
                .frame(width: 24, height: 24)
        }
    }

    /// "42:10 · Bench press", or while resting "Rest 1:12 · Squat" with the countdown in ember.
    private func detail(_ state: WorkoutMiniBarState) -> some View {
        HStack(spacing: 0) {
            Group {
                if let remaining = state.restRemaining {
                    Text("timer.rest")
                    Text(verbatim: " ")
                    Text(Fmt.clock(remaining)).foregroundStyle(NT.Colors.ember)
                } else {
                    Text(state.elapsed)
                }
            }
            .layoutPriority(1)
            if !state.exerciseName.isEmpty {
                Text(verbatim: " · \(state.exerciseName)")
            }
        }
        .font(NT.Fonts.footnote).foregroundStyle(NT.Colors.ink2).tabular().lineLimit(1)
    }

    private func accessibilityValue(_ state: WorkoutMiniBarState) -> Text {
        guard let remaining = state.restRemaining else { return Text(verbatim: state.exerciseName) }
        let rest = "\(Fmt.localized("timer.rest")) \(Fmt.clock(remaining))"
        return Text(verbatim: state.exerciseName.isEmpty ? rest : "\(rest), \(state.exerciseName)")
    }

    static let height: CGFloat = NT.Size.cardButton
}

/// Pins the bar above the tab bar via a bottom safe-area inset, so the tab's own bottom insets (Fuel's add bar)
/// stack above it. Stays out of the way while the keyboard is up, like a bar attached to the tab bar would.
private struct WorkoutMiniBarInset: ViewModifier {
    var isEnabled: Bool
    @State private var isKeyboardUp = false

    func body(content: Content) -> some View {
        content
            .safeAreaInset(edge: .bottom, spacing: 0) {
                if isEnabled && !isKeyboardUp {
                    WorkoutMiniBar()
                        .padding(.horizontal, NT.Spacing.screenH)
                        .padding(.bottom, 8)
                        .transition(.move(edge: .bottom).combined(with: .opacity))
                }
            }
            .animation(.spring(response: 0.35, dampingFraction: 0.85), value: isEnabled)
            .onReceive(NotificationCenter.default.publisher(for: UIResponder.keyboardWillShowNotification)) { _ in
                isKeyboardUp = true
            }
            .onReceive(NotificationCenter.default.publisher(for: UIResponder.keyboardWillHideNotification)) { _ in
                isKeyboardUp = false
            }
    }
}

extension View {
    /// The workout mini bar on a tab. Apply to each tab's root content, unconditionally (`isEnabled` shows and hides it).
    ///
    /// iOS 26's `tabViewBottomAccessory(isEnabled:)` would be the native container, but the iOS 26.1 SDK in
    /// Xcode 26.1.1 does not declare that overload, and the 26.0 one leaves an empty glass capsule on 26.1 when
    /// hidden, so every iOS version uses this inset (glass on iOS 26, a surface capsule before).
    func workoutMiniBar(isEnabled: Bool) -> some View {
        modifier(WorkoutMiniBarInset(isEnabled: isEnabled))
    }

    @ViewBuilder
    fileprivate func miniBarChrome() -> some View {
        if #available(iOS 26.0, *) {
            glassEffect(.regular.interactive(), in: Capsule())
        } else {
            background(NT.Colors.surface, in: Capsule())
                .overlay(Capsule().strokeBorder(NT.Colors.hairline, lineWidth: 1))
                .shadow(color: .black.opacity(0.5), radius: 12, y: 8)
        }
    }
}
